Add transformation test make target and docs (#1722)
GoogleCloudPlatform/ops-agent: dev-docs/transformation-tests.md

# Transformation Tests

> The transformation tests are tests that will run Fluent Bit and OpenTelemetry Operations Collector (otelopscol) directly in your environment and test the results of various logging operations to stdout. This is to check the correctness of log parsing expectations in a more lightweight way than full integration tests.

Reviewer: "test… to stdout" sounds a bit off. Did you mean "test… *via* stdout" (i.e., "…by printing to stdout")?

## Makefile Usage

NOTE: It is assumed that you have followed the [makefile setup instructions](./makefile.md#usage).

For your convenience, there is a Makefile target to run transformation tests:
```bash
make transformation_test
```
You can also update the transformation test goldens:
```bash
make transformation_test_update
```
These targets are part of the `precommit` and `precommit_update` targets respectively.

Read on if you are interested in how the transformation tests actually work.

## Usage

The transformation tests take paths to Fluent Bit and otelopscol binaries to execute directly and compare expected output. To run the tests, first you will need to build both binaries. There are scripts to locally build each agent:
```bash
sh ./builds/fluent_bit_local.sh

> sh ./builds/otel_local.sh

Reviewer: Should we add Makefile targets for these? *And* make `transformation_test…` targets dependent on them?

> sh ./builds/otel_local.sh

Author: Done in [b94c571](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/b94c571b964765d12a384bd60d0f47ef8bc99c58)

```
After running each script, you will have a `dist` folder that contains each binary.

The paths to each can passed into the test using environment variables or flags.
```bash

> go test ./transformation_test -flb=$(pwd)/dist/fluent_bit -otelopscol=$(pwd)/dist/otelopscol

Author: Done in [b94c571](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/b94c571b964765d12a384bd60d0f47ef8bc99c58)

```
or
```bash
FLB=$(pwd)/dist/fluent_bit OTELOPSCOL=$(pwd)/dist/otelopscol go test ./transformation_test
```

If you have updated the Fluent Bit or otelopscol submodules and need to update the golden testdata, you can add the `-update` flag to the test.
```bash
FLB=$(pwd)/dist/fluent_bit OTELOPSCOL=$(pwd)/dist/otelopscol go test ./transformation_test -update
```

> (Note: this is a flag to the test, not to `go test`. The flag must go after the specified package, not before.)

Author: Done in [b94c571](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/b94c571b964765d12a384bd60d0f47ef8bc99c58)

## Usage

The transformation tests take paths to Fluent Bit and otelopscol binaries to execute directly and compare expected output. To run the tests, first you will need to build both binaries. Use the build scripts to build each agent and install them into the `dist` folder:
```bash
sh ./builds/fluent_bit.sh $(pwd)/dist
SKIP_OTEL_JAVA=true sh ./builds/otel.sh $(pwd)/dist
```
After running each script, you will have a `dist` folder that contains both agents within the full install directory. (Note that these instructions will not build OTel Java, if that is needed do not pass `SKIP_OTEL_JAVA`).

The paths to each can passed into the test using environment variables or flags.
```bash
go test ./transformation_test \
-flb=${PWD}/dist/opt/google-cloud-ops-agent/subagents/fluent-bit/bin/fluent-bit \
-otelopscol=${PWD}/dist/opt/google-cloud-ops-agent/subagents/opentelemetry-collector/otelopscol
```
or
```bash
FLB=${PWD}/dist/opt/google-cloud-ops-agent/subagents/fluent-bit/bin/fluent-bit \
OTELOPSCOL=${PWD}/dist/opt/google-cloud-ops-agent/subagents/opentelemetry-collector/otelopscol \
go test ./transformation_test
```

> If you have updated the Fluent Bit or otelopscol submodules and need to update the golden testdata, you can add the `-update` flag to the test.

Author: Fixed in [6e0c84a](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/6e0c84a9646afc03f5b0e4545fccfad1b60183a0)

```bash
FLB=${PWD}/dist/opt/google-cloud-ops-agent/subagents/fluent-bit/bin/fluent-bit \
OTELOPSCOL=${PWD}/dist/opt/google-cloud-ops-agent/subagents/opentelemetry-collector/otelopscol \
go test ./transformation_test -update
```

> (Note: the`-update` flag belongs to the test, not to `go test`. The flag must go after the specified package, not before.)

Author: Fixed in [6e0c84a](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/6e0c84a9646afc03f5b0e4545fccfad1b60183a0)

```
These targets are part of the `precommit` and `precommit_update` targets respectively.

If you are on a new branch that updates Fluent Bit or otelopscol and are locally verifying the transformation tests, you can force the agents to rebuild by deleting the `dist` folder before running `make transfomration_test`; the target will rebuild the agents if they are not present.

Read on if you are interested in how the transformation tests actually work.

## Usage

The transformation tests take paths to Fluent Bit and otelopscol binaries to execute directly and compare expected output. To run the tests, first you will need to build both binaries. Use the build scripts to build each agent and install them into the `dist` folder:
```bash
sh ./builds/fluent_bit.sh $(pwd)/dist
SKIP_OTEL_JAVA=true sh ./builds/otel.sh $(pwd)/dist
```

> After running each script, you will have a `dist` folder that contains the full installation of both agents.

Reviewer: Not quite the full installation, since the OTel Java stuff would be missing… How about this:
```suggestion
After running each script, you will have a `dist` folder that contains the agent binaries at installation paths.
```

Author: Done in a similar way in [0db4c79](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/0db4c799758b739846eb41852b0ef9654fe4f9c5)

sh ./builds/fluent_bit.sh $(pwd)/dist
SKIP_OTEL_JAVA=true sh ./builds/otel.sh $(pwd)/dist
```

> After running each script, you will have a `dist` folder that contains both agents within the full install directory. (Note that these instructions will not build OTel Java, if that is needed do not pass `SKIP_OTEL_JAVA`).

Author: Fixed in [6e0c84a](https://github.com/GoogleCloudPlatform/ops-agent/pull/1722/commits/6e0c84a9646afc03f5b0e4545fccfad1b60183a0)
